lib/deep_cover/custom_requirer.rb

```diff
@@ -43,96 +41,162 @@ def compute_subset
-    # Throws :use_fallback in case caller should delegate to the default #require.
-    # Reasons given could be:
+    # Calls &fallback_block with the reason as parameter if the work couldn't be done.
+    # The possible reasons are:
     #  - :not_found if the file couldn't be found.
+    #  - :not_in_covered_paths if the file is not in the paths to cover
     #  - :cover_failed if DeepCover couldn't apply instrumentation the file found.
-    #  - :not_supported for files that are not supported (such as ike .so files)
+    #  - :not_supported for files that are not supported (such as .so files)
     #  - :skipped if the filter block returned `true`
-    # Exceptions raised by the required code bubble up as normal.
-    #     It is *NOT* recommended to simply delegate to the default #require, since it
-    #     might not be safe to run part of the code again.
-    def require(path)
+    # Exceptions raised by the required code bubble up as normal, except for
+    # SyntaxError, which is turned into a :cover_failed which calls the fallback_block.
+    def require(path, &fallback_block)
       path = path.to_s
-      ext = File.extname(path)
-      throw :use_fallback, :not_supported if ext == '.so'
-      path += '.rb' if ext != '.rb'
-      return false if @loaded_features.include?(path)
 
       found_path = resolve_path(path)
 
-      throw :use_fallback, :not_found unless found_path
-      return false if @loaded_features.include?(found_path)
+      if found_path
+        return false if @loaded_features.include?(found_path)
+        return false if @paths_being_required.include?(found_path)
+      end
+
+      DeepCover.autoload_tracker.wrap_require(path, found_path) do
+        # Either a problem with resolve_path, or a gem that will be added to the load_path by RubyGems
+        return yield(:not_found) unless found_path
 
-      throw :use_fallback, :skipped if filter && filter.call(found_path)
+        begin
+          @paths_being_required.add(found_path)
+          return yield(:not_in_covered_paths) unless @load_paths_subset.within_lookup?(found_path)
+          return yield(:not_supported) if found_path.end_with?('.so')
+          return yield(:skipped) if filter && filter.call(found_path)
 
-      cover_and_execute(found_path)
+          cover_and_execute(found_path) { |reason| return yield(reason) }
 
-      @loaded_features << found_path
+          @loaded_features << found_path
+        ensure
+          @paths_being_required.delete(found_path)
+        end
+      end
       true
     end
 
+    ### Not currently used ###
     # Homemade #load to be able to instrument the code before it gets executed.
     # Note, this doesn't support the `wrap` parameter that ruby's #load has.
     # Same return/throw as CustomRequirer#require, except:
     # Cannot return false since #load doesn't care about a file already being executed.
-    def load(path)
-      found_path = resolve_path(path)
+    def load(path, &fallback_block)
```

Comment:
I'd avoid capturing blocks when not needed.

Reply:
As discussed, replaced with comment to keep the information available near the signature.